python/dgl/nn/pytorch/link.py

```diff
@@ -0,0 +1,128 @@
+"""Torch modules for link prediction."""
+# pylint: disable= no-member, arguments-differ, invalid-name, W0235
+import torch.nn as nn
+import torch.nn.functional as F
+
+from ... import function as fn
+
+__all__ = ['HadamardPredictor']
+
+class HadamardPredictor(nn.Module):
+    r"""
+
+    Description
+    -----------
+    Link predictor based on Hadamard product (element-wise product) introduced in
+    `Open Graph Benchmark: Datasets for Machine Learning on Graphs
+    <https://arxiv.org/abs/2005.00687>`. It applies the Hadamard product to pairs
+    of node representations and passes the results to an MLP for the final prediction.
+
+    Parameters
+    ----------
+    in_feats : int
+        Input feature size.
+    hidden_feats : int
+        Hidden feature size.
+    out_feats : int
+        Output feature size.
+    num_layers : int, optional
+        Number of linear layers, which should be at least 2.
+    dropout : float, optional
+        Dropout to apply in MLP.
+    activation : callable activation function/layer or None, optional
+        If not None, applies an activation function in MLP.
+
+    Examples
+    --------
+    >>> import dgl
+    >>> import torch
+    >>> from dgl.nn import HadamardPredictor
+
+    Case1: score the edges of a graph
+
+    >>> g = dgl.graph(([0, 1, 2], [1, 2, 3]))
+    >>> in_feats = 2
+    >>> feat = torch.randn((g.num_nodes(), in_feats))
+    >>> predictor = HadamardPredictor(in_feats=in_feats, hidden_feats=3, out_feats=1)
+    >>> predictor(feat, g).shape
+    torch.Size([3, 1])
+
+    Case2: score arbitrary node pairs
+
+    >>> num_pairs = 3
+    >>> feat_i = torch.randn((num_pairs, in_feats))
+    >>> feat_j = torch.randn((num_pairs, in_feats))
+    >>> predictor = HadamardPredictor(in_feats=in_feats, hidden_feats=3, out_feats=1)
+    >>> predictor((feat_i, feat_j)).shape
+    torch.Size([3, 1])
+    """
+    def __init__(self,
+                 in_feats,
+                 hidden_feats,
+                 out_feats,
+                 num_layers=2,
+                 dropout=0.,
+                 activation=F.relu):
```

Comment:
Why not let the users pass in his/her own MLP module instead?  Something like
```
HadamardPredictor(
    nn.Sequential(
        nn.Linear(in_feats, hid_feats),
        nn.Dropout(0.5),
        nn.Linear(hid_feats, out_feats)))
```
This way could give them more flexibility like GIN.

Reply:
In that case, I wonder if it will be better to simply have this function compute the Hadamard product. Since the MLP is applied at the end, we may not need to wrap an MLP this class.

Reply:
That depends on what a predictor will produce.  If we say a predictor is "something that takes in the incident nodes' representations and output a single scalar for each edge", then wrapping an MLP still makes sense.

Reply:
We can rename `HadamardPredictor` to `HadamardProduct` in that case, just like `DotProduct`, `CosineDistance`, etc. Meanwhile, this is the same case as `EdgeMLP`, which concatenates pairs of node representations and passes the results to an MLP. If we do not include an MLP, is it still beneficial to have a function/class instance that concatenates pairs of node representations?